FGEGraphics/UISystem/UIInputLabel.cs

```diff
@@ -0,0 +1,469 @@
+        {
+            Internal.CursorLeft = Internal.CursorRight;
+        }
+        UpdateText();
+    }
+
+    /// <summary>Modifies the current selection based on mouse clicks/drags.</summary>
+    /// <param name="shiftDown">Whether the shift key is being held.</param>
+    public void TickMouse(bool shiftDown)
+    {
+        if (!MouseDown || Internal.TextChain.Count == 0)
+        {
+            return;
+        }
+        float relMouseX = Window.MouseX - X;
+        float relMouseY = Window.MouseY - Y;
+        UIElementText.ChainPiece lastPiece = Internal.TextChain[^1];
+        if (lastPiece.YOffset + (lastPiece.Font.FontDefault.Height * lastPiece.Text.Lines.Length) < relMouseY)
+        {
+            TickMousePosition(TextContent.Length, shiftDown);
+            return;
+        }
+        int indexOffset = 0;
+        foreach (UIElementText.ChainPiece piece in Internal.TextChain)
+        {
+            for (int i = 0; i < piece.Text.Lines.Length; i++)
+            {
+                RenderableTextLine line = piece.Text.Lines[i];
+                string content = line.ToString();
+                if (piece.YOffset + (i + 1) * piece.Font.FontDefault.Height >= relMouseY)
+                {
+                    if (line.Width < relMouseX)
+                    {
+                        TickMousePosition(indexOffset + content.Length, shiftDown);
+                        return;
+                    }
+                    float lastWidth = 0;
+                    for (int j = 0; j <= content.Length; j++)
+                    {
+                        float width = piece.Font.MeasureFancyText(content[..j]);
+                        if (width >= relMouseX)
+                        {
+                            int diff = relMouseX - lastWidth >= width - relMouseX ? 0 : 1;
+                            TickMousePosition(indexOffset + j - diff, shiftDown);
+                            return;
+                        }
+                        lastWidth = width;
+                    }
+                }
+                indexOffset += content.Length;
+            }
+            indexOffset++;
+        }
+    }
+
+    /// <summary>Handles various control key combinations.</summary>
+    /// <param name="keys">The current keyboard state.</param>
+    public void TickControlKeys(KeyHandlerState keys)
+    {
+        if (keys.CopyPressed && Internal.IsSelection)
+        {
+            TextCopy.ClipboardService.SetText(Internal.TextBetween.Content);
+        }
+        if (keys.AllPressed)
+        {
+            Internal.CursorLeft = 0;
+            Internal.CursorRight = TextContent.Length;
+            UpdateText();
+        }
+    }
+
+    /// <inheritdoc/>
+    public override void Tick(double delta)
+    {
+        base.Tick(delta);
+        if (!Selected)
+        {
+            return;
+        }
+        KeyHandlerState keys = Window.Keyboard.BuildingState;
+        if (keys.Escaped)
+        {
+            Closed?.Invoke();
```

Comment:
presumably should `Selected = false;` here as well

Reply:
The `HandleClose` method, which does that, is appended to the `Closed` event in the constructor